### ISSUE TITLE
Incorrect crop size in mmseg radio

### ISSUE THREAD

**Issue by collinmccarthy** [NVlabs/RADIO#73]
Hello,

Looks like the crop size is incorrect in the RADIO ADE20K config. 

It's listed as (518, 518) [here](https://github.com/NVlabs/RADIO/blob/6c2dd67fd7d5a7c2b4e145be5755da9f573c33fa/mmseg/configs/radio/radio_linear_8xb2-80k_ade20k-512x512.py#L10) but it should be (512, 512) like it is in E-RADIO and the ADE20K base config. 

Thanks,
-Collin

**Comment by gheinrich** [comment]
Hello, for RADIO the crop size was set to 518 as this was the nearest multiple of the RADIOv1 patch size of 14.
However now that we have released the patch-size-16 RADIOv2 you're right that we should update the MMSEG crop size to 512.

**Comment by mranzinger** [comment]
Btw, have you seen that we just released RADIOv2.5 ViT-B and L models? If not, check out the tech report. They may very well work better for your use case.

**Comment by collinmccarthy** [comment]
@mranzinger Yes I did see that, very excited to try them out, thanks!

**Comment by collinmccarthy** [comment]
@mranzinger Is there a way to specify what version I want from HF? Specifically, what do I pass as the `repo_id` for `AutoModel.from_pretrained(repo_id)` to specify each of the following versions in `radio.common.RESOURCE_MAP`? 

- Version "radio_v2.5-b" $\rightarrow$ `repo_id="nvidia/RADIO-B"` ?
- Version "radio_v2.5-l" $\rightarrow$ `repo_id="nvidia/RADIO-L"` ?
- Version "radio_v2.1" $\rightarrow$ ??
- Version "re-radio_v2" $\rightarrow$ ??

Or do I have to use torchhub to specify these versions (and therefore modify `radio.mmseg.radio.py` to use this API instead)?

**Comment by mranzinger** [comment]
@gheinrich is the expert on the HFHub code. But I believe you want https://huggingface.co/nvidia/RADIO-L, so yes, I think you're right with your `repo_id` map. For `radio_v2.1 -> repo_id="nvidia/RADIO"` and `e-radio_v2 -> repo_id="nvidia/E-RADIO"`.

I may be ignorant about all of the features of HFHub, but I will say that it's been easier to deal with model versions and releases using TorchHub.

**Comment by collinmccarthy** [comment]
Okay great, thanks. I'll give this a shot and let you know if it works.

**Comment by collinmccarthy** [comment]
The mapping below appears to be correct (the same one discussed above). It would be nice if this is in `radio.common.RESOURCE_MAP` for future versions, but not a huge deal.

```python
HF_RESOURCE_MAP = {  # Version key in RESOURCE_MAP to HuggingFace repo id
    # RADIOv2.5
    "radio_v2.5-b": "NVIDIA/RADIO-B",
    "radio_v2.5-l": "NVIDIA/RADIO-L",
    # RADIO
    "radio_v2.1": "NVIDIA/RADIO",
    "radio_v2": None,
    "radio_v1": None,
    # E-RADIO
    "e-radio_v2": "NVIDIA/E-RADIO",
}
```

However, it looks like the HF vs. torchhub models for RADIO v2.1 and E-RADIO v2.0 have some slight differences. Not sure if this is expected or not, or if it's a big deal for anyone and worth fixing, but it's failing some tests that I have (as well as `test_hf.py`) so I thought I'd share.

For RADIO v2.1, the HF model at `NVIDIA/RADIO`  has `summary_idxs=tensor([0,2])` while the torchhub model has `summary_idxs=tensor([0,1,2,3])`. For my application this is fine, but this fails the `test_hf.py` test because the summary tensor is different. But the features output tensor and all other buffers/parameters/submodules are the same. 

For E-RADIO v2.0, the HF model at `NVIDIA/E-RADIO` has `summary_idxs=tensor([0,1,2])` while the torchhub model has `summary_idxs=tensor([0,1,2,3])` again. It also looks like the relative_bias at `model.levels.N.blocks.M.attention_blocks.K.attn.pos_emb_funct.relative_bias` is slightly different. The values are extremely close, e.g. 6.2875 to 6.2899, but I'm not sure why they're different or if this matters. I am struggling to reproduce the E-RADIO ade20k results, so it could be related to this, but my guess is these are close enough that it doesn't matter too much. I'll try switching to the torchhub version and see if it fixes it. 

I'm not sure if these things were this way before the v2.5 push either, I wasn't testing things in this way before.

**Comment by mranzinger** [comment]
The `summary_idxs` tensor is correct for the HF model actually. I'll need to look into why you're getting `[0, 1, 2, 3]` for TorchHub. I assume you've run `force_reload=True` recently since you've already fetched the v2.5 models, right? I thought I fixed it, but I suppose I'll need to check again 😅

How are you using the `HF_RESOURCE_MAP` you're showing? Is it just for reference, or are you using as part of an API that we should consider upstreaming into main?

For E-RADIO, how different are your ADE20k results versus our reported?

**Comment by collinmccarthy** [comment]
For the torchub models I'm directly calling `torch.hub.load_state_dict_from_url` using the url in the resource map.  So any issue with the two torchhub models I mentioned is with the models at those URLs specifically. When I used `test_hf.py` just to see if that was any different (it wasn't, same discrepancies), I just manually deleted my `~/.cache/torch/hub/<radio>` directory because it wasn't overwriting everything from scratch when just specifying `force_reload=True`. 

I'm using the `HF_RESOURCE_MAP` as part of a script I wrote to generate keyword arguments for a version of RADIO that doesn't rely on pretrained checkpoints to create the model. The script compares my version to the torchhub and HF versions, and checks for any mismatched buffers/parameters/submodules. Basically I needed (maybe 'wanted' is a better word) to build RADIO models from just a config file with plain data types, and then separately load the weights using just `load_state_dict_from_url`. Constructing the models this way allows me to create model wrappers or extend the model classes in the easiest and most readable way. This is mostly beneficial when fine-tuning RADIO, especially when trying to integrate it into things like mmdetection where managing a ton of config files needs to be done carefully, and using explicit keyword arguments helps a lot. 

My best results on ADE20K for E-RADIO were 22.7 mIoU. I don't have 8 GPUs so I'm using 2 GPUs w/ BS=8. I tried forcing `sync_bn=torch` (best mIoU) and leaving the default `sync_bn=None` (best aAcc), using both fp32 (works) and amp fp16 (didn't work at all). In comparison, I got 50.72 mIoU for RADIO v2.1, and the results were almost identical for fp32, amp w/ fp16, with or without sync_bn (which makes sense since I think only the head uses BN and it's already configured as SyncBN). I'm training RADIO v2.5-B and v2.5-L and it looks like v2.5-L is definitely better than v2.1. Too early to tell on v2.5-B, but it looks to be very good and I'm very impressed with how close it is given it has less than 100M parameters.

**Comment by collinmccarthy** [comment]
I forgot, I made two other changes to the E-RADIO config. Using `test_cfg=dict(mode='slide')` was timing out in the last few val iterations on GPU 0, leading to some weird issues with broadcasting the tmpdir dir eval (similar to [this mmengine issue](https://github.com/open-mmlab/mmengine/issues/1178)) . Using `mode='whole'` fixed this but I had to pad the image to multiples of 32. I've attached the config file I had to use (with a dummy .txt extension for GitHub) where the only places I've updated it are marked with `# UPDATED`. 

[eradio_linear_8xb2-80k_ade20k-512x512.py.txt](https://github.com/user-attachments/files/16385539/eradio_linear_8xb2-80k_ade20k-512x512.py.txt)

**Comment by gheinrich** [comment]
Hello Collin, in some earlier experiments I ran, I had found "slide" to work better than "whole":
![image](https://github.com/user-attachments/assets/7d4b41e8-3a6c-48ea-9643-953d95ebbe18)

That alone should not explain the discrepancy you're seeing.

I understand you are instantiating the model and loading the weights manually, is this correct? Can I check with you that the E-RADIO model you are instantiating is a "window size 16"? The window size can have a dramatic impact on mIoU, I found 16 to be optimal. For example I would see a boost of +9 on mIoU going from window size 7 to 16 (without re-training).

If that doesn't help, I suggest looking at inference visualizations, you can enable them with something like this in your config:

```
default_hooks = dict(
    visualization=dict(type='SegVisualizationHook',  draw=True, interval=20))
```

**Comment by collinmccarthy** [comment]
Hi Greg, very interesting, thanks for sharing! I'll have to look into the "slide" implementation in more detail, I just assumed it would crop/stitch the output to give exactly the same results. I'm not sure why it was causing issues for me, after I figured out that "whole" fixed the issue I just moved on.

For all of the results I've discussed so far, I'm using the exact same mmseg setup as in this codebase. All I've changed is the config file I posted above to fix the issue, and switched to using 2x GPUs with a batch size of 8. I've attached the config I get from the following

```python
hf_model = AutoModel.from_pretrained("NVIDIA/E-RADIO", trust_remote_code=True)
print(pprint.pformat(hf_model .config, sort_dicts=False))
```

[eradio_config.txt](https://github.com/user-attachments/files/16394484/eradio_config.txt)

It looks like the config has `vitdet_window_size: None`. On top of that, `VitDetArgs` is imported in `radio.hf_model.py` but never used. Also, the torchhub checkpoint at 'https://huggingface.co/nvidia/RADIO/resolve/main/eradio_v2.pth.tar?download=true' does not have `vitdet_window_size` in checkpoint['args'], so it looks like that needs to be manually passed into hubconf.radio_model() or added afterwards via `apply_vitdet_arch()`. 

I'll try training some models with my torchhub-based API while passing in vitdet_window_size=16 and see if that fixes it. Thanks!

**Comment by collinmccarthy** [comment]
Sorry, I see what's happening here with `vitdet_window_size`. ViTDetArgs is only used for the normal ViT models, not E-RADIO, and the window size is already set to 16 during the timm model registration, e.g. with `create_model(model='eradio')`. 

One more hint as to what might be going on. When calling `_, features = self.base_model(x)` [here](https://github.com/NVlabs/RADIO/blob/8e5b42ba020c2cb8a76f2c63f24ce092b9b07caa/mmseg/radio.py#L60) the features returned always have shape `B, N, C` and need to be re-shaped. I used `self.patch_size` which is 16 for E-RADIO which produced the same feature vector as directly calling `ERADIO.forward_features()`. But given that I had to add this code, maybe something else changed in that forward_features() implementation, like I shouldn't actually be taking the norm, or I shouldn't be using the high-res neck, or something. 

So I'm not sure what the issue is with E-RADIO and mmseg. Sorry I can't narrow down the issue more. I'll stick with RADIO v2.5-B as my "efficient" model for now and if you want me to submit a PR just so you can see the changes I've made so far to try and get it to work, I'm happy to do that. Thanks.

**Comment by mranzinger** [comment]
PRs are always welcome. The ViT-B model indeed seems to be really high quality, so I hope you can make some progress with it. I'm actually generally interested in what you're up to if you'd be willing to shoot me and/or Greg an email with details, if you're interested.

**Comment by collinmccarthy** [comment]
Sounds great, I'll email you and Greg to sync up on how I'm using this and we can discuss the PR then. Thanks!

**Comment by Saloni1Parekh609** [comment]
@mranzinger This might be a silly question but I was wondering what the hyperparameters for SemSeg on ADE20k look like if the pape rmentions 160k steps but the config file is made for 80k steps? Additionally, are these hyperparameters the same for all the models in table 1 in AMRadio?

`param_scheduler = [
    dict(type="LinearLR", start_factor=1e-6, by_epoch=False, begin=0, end=1500),
    dict(
        type="PolyLR",
        eta_min=0.0,
        power=1.0,
        begin=1500,
        end=80000,
        by_epoch=False,
    ),
]`

**Comment by mranzinger** [comment]
Yes, we use the same learning rate schedule across the board. The only difference between the 80k and 160k spec is the duration.

**Comment by 836469383** [comment]
Sorry to bother. When I reproduced mmseg by Radio-B, I changed the crop size to 512, in_channels and channels to 768. It reported an error like the fig when it iterated to 5000. Do you have any suggestions? Any help will be thankful. 

![Image](https://github.com/user-attachments/assets/1b4ad9b2-a623-42d9-8859-fb5410ec7dea)

**Comment by 836469383** [comment]
Oh! It works. Thanks!!
I truly admire the coding skills. If u have the time, could u share some advice on how i can reach the level!(★ ω ★)
